File: Algorithm/2024.01/wolfy916/T 제거하기.py
'''
T 제거하기 - 코드트리 골드5
분류 : 스택
'''
import sys

# [A] 입력함수 초기화
def input():
    return sys.stdin.readline().rstrip('\n')

# [Main]
if __name__ == "__main__":
    S = input()
    T = input()

    N, M = len(S), len(T)
    stack = []
    p1, p2 = -1, M - 1
    for i in range(N):
        stack.append(S[i])
        p1 += 1
        flag = False
        while p1 >= 0 and p2 >= 0:
            if stack[p1] != T[p2]: break
            p1 -= 1
            p2 -= 1
            if p2 < 0: flag = True
        
        if flag:
            for _ in range(M):
                stack.pop()

        p1 = len(stack) - 1
        p2 = M - 1

    print("".join(stack))

File: Algorithm/2024.01/PakaOxO/T제거하기.js
/**
 * CodeTree_T 제거하기
 *  - 문제 분류: 문자열, 스택
 */
const solution = () => {
  /* 변수 관리 */
  const [S, T] = require("fs").readFileSync("./dev/stdin/T제거하기.txt").toString().trim().split("\n");
  const stack = [];
  let top = -1;
  const tLen = T.length;
  let subPointer = tLen - 1;

  /* 메인 로직 */
  for (let i = 0; i < S.length; i++) {
    const char = S.charAt(i);
    stack.push(char);
    top++;

    let flag = false;
    while (top >= 0 && subPointer >= 0) {
      if (stack[top] !== T.charAt(subPointer)) break;
      subPointer--;
      top--;

      if (subPointer < 0) flag = true;
    }

    if (flag) {
      for (let j = 0; j < tLen; j++) stack.pop();
    }
    top = stack.length - 1;
    subPointer = tLen - 1;
  }

  /* 정답 반환 */
  return stack.join("");
};

console.log(solution());
